docs/moderation: document moderation section (#15)
botlabs-gg/yagpdb-docs-v2: content/moderation/moderation-overview.md

+++
archetype = "default"
title = "Moderation"
weight = 1
+++

Everything in moderation, including moderation.

<!--more-->

![Overview of the moderation page](overview_moderation.png)

<center>

**1** Mod Log Channel **2** Report Channel **3** Moderation DM Error Channel **4** Clean Command **5** Report
Command **6** Unban Logging **7** Giverole/Takerole commands **8** Give/Takerole Permitted Roles **9**
Give/Takerole Logging **10** External Ban Logging **11** External Kick Logging **12** External Timeout Logging **13**
Moderation Tabs **14** Bot Permissions Notice

</center>

## Moderation Page

The moderation page allows configuration of common settings.

### Mod Channels

> 1. **Mod Log** - The mod log is the channel YAGPDB will send log messages to when most moderation commands are used. It
> includes username and user ID of both the moderator, and the moderated user. It additionally includes the reason for

Reviewer: ```suggestion
1. **Mod Log** - The mod log is the channel YAGPDB will send log messages to when most moderation commands are used. Log entries
include username and user ID of both the moderator, and the moderated user. They additionally include the reason for
```

relevant commands (defaults to `<no reason provided>` when not provided), and a [log](/moderation/logging) of the
channel where the command was run.

> 2. **Report** - The channel for YAGPDB to send report messages to when users use the report command. Similar to the mod
> log, it includes username and user ID of the user reporting, and the user reported, as well as the reason and message
> log of the channel.

Reviewer: ```suggestion
2. **Report** - The channel for YAGPDB to send report messages to when users use the report command. Similar to the mod
log, report entries include username and user ID of the user reporting, and the user reported, as well as the reason and message
log of the channel.
```

3. **Moderation DM Error** - The channel where YAGPDB will send error messages from executing template script within
[moderation DMs](#moderation-dms).

### Command Toggles

Certain potentially destructive moderation commands are disabled by default. These commands are:

- The **clean command** (**4**) toggles the [clean command](/commands/all-commands#clean).
- The **report command** (**5**) toggles the [report command](/commands/all-commands#report).
- **Giverole/Takerole commands** (**7**) toggle the [giverole](/commands/all-commands#giverole) and
[takerole](/commands/all-commands#takerole) commands.

### Unban/Unmute Logging

When the `unban`/`unmute` commands are used, YAGPDB will log their usage to the [mod log](#mod-channels) similarly to

> the other moderation commands, except without the logs.

Reviewer: I am confused by the phrase "except without the logs" here.

Reviewer 2: agreed, can omit

> When using the `ban` and `mute` commands with a duration, YAGPDB will essentially execute the inverse commands as itself
> at the end of the duration. i.e, if you execute `ban @user 1d`, YAGPDB will ban the user, and 1 day later YAGPDB will
> execute the `unban` command with reason "Timed ban expired." This action appears in the mod log as YAGPDB being the
> moderator.

Reviewer: ```suggestion
When using the `ban` and `mute` commands with a duration, YAGPDB will automatically undo the punishment after the specified duration. For instance, if you execute `ban @user 1d`, YAGPDB will ban the user, and 1 day later YAGPDB will
execute the `unban` command with reason "Timed ban expired." This action appears in the mod log as YAGPDB being the
moderator.
```

I think the phrase "execute the inverse commands as itself" is not layman-friendly.

Reviewer 2: me when i make no sense

Toggling Unban Logging (**6**) allows you to enable or disable the logging of Unmute events.

Toggling Give/Takerole Logging (**9**) allows you to enable or disable the logging of Give/Takerole events.

### External Logging

> By default, YAGPDB will only log actions it takes, i.e. if you kick a user with the bot's `kick` command, it will log
> it, but not if you kick the user directly via the user context menu. Enabling external logging for different moderation

Reviewer: ```suggestion
By default, only actions performed through YAGPDB will be recorded in the mod-log. Hence, kicking a user using YAGPDB's `kick` command will result in a log entry, whereas kicking the user directly through Discord's context menu will not be recorded. Enabling external logging for different moderation
```

actions (**10**), (**11**), (**12**) allows the bot to use the server Audit Log to retrieve moderation actions taken by
other bots or other users.

YAGPDB must have View Audit Log permissions to use this feature.

## Permissions Notice

> The moderation plugin, among other plugins, requires additional permissions to function properly. The notice at the top
> of the screen (**14**) will inform you of the required permissions for the plugin and alert you if the bot is missing
> permissions.

Reviewer: ```suggestion
The moderation plugin requires additional permissions to function properly. The notice at the top
of the screen (**14**) will inform you of the required permissions for the plugin and alert you if the bot is missing
permissions.
```
superfluous

members for the kick command, etc.) are able to run the corresponding moderation command. Selecting permitted roles
(**1**) will allow said roles to run the moderation commands **even if** they do not have required Discord permissions.

Give/Takerole Permitted Roles (**8**) operate in the same way.

#### Reason Optional

By default, Timeout, (Un)Mute, Kick, and Ban tools require the moderator to specify a reason. This forces your staff
(and yourself, no exceptions!) to specify a reason for taking mod action for the mod logs, and also by default to be
DM'd to the offending user.

To remove the requirement and make the reason optional, you may toggle this on (**2**).

#### Moderation DMs

Moderation DMs are template scripts (**3**) which are executed within the offending user's DM. Text output to the response will
be directly sent to the offending user.

{{% button href="/reference/templates" style="transparent" %}}Templates Reference{{% /button %}}

Additional template dot context data is available for Moderation Actions:

- `{{.Reason}}`: The reason specified in the timeout.
- `{{.Author}}`: The author of the punishment, is a [user object](/reference/templates#user).

> - `{{.Duration}}`: The duration of the mod action in [time.Duration format](/reference/templates#time).

Reviewer: ```suggestion
- `{{.Duration}}`: The duration of the mod action as a [time.Duration](/reference/templates#time).
```

- `{{.HumanDuration}}`: The duration in a human friendly format (`1 hour and 3 minutes` for example).
- `{{.WarningID}}`: The ID of the warning (when using the warn command).

Any errors which occur in the template execution will not be logged by default. The offending user will only receive a
DM from your server saying **Failed executing template.** Setting a [Moderation DM Error channel](#mod-channels) allows
for the more detailed errors to be captured.

Moderation actions which send DMs are:

- Timeout
- Mute
- Kick
- Ban
- Warn

### Timeout

> Timing out a member employs Discord's timeout feature to prevent users from interacting with your server. Discord's
> timeout temporarily removes the member's permissions to send messages, add reactions, and join voice channels,
> indiscriminantly and regardless of roles or overrides. Using a bot to timeout a member allows you to assign a specific

Reviewer: ```suggestion
Timing out a member employs Discord's timeout feature to prevent users from interacting with your server. Discord's
timeout temporarily removes the member's permissions to send messages, add reactions, and join voice channels,
regardless of roles or overrides. Using a bot to timeout a member allows you to assign a specific
```

Reviewer 2: this part needs to be changed anyways, as i realized ~~when lotnik timed me out~~ that you lose more than just these three permissions. I'm struggling to find Discord documentation about timeout but my working hypothesis is all permissions except read and message history are removed.

Reviewer 3: > Timed out members will temporarily lose all permissions except `VIEW_CHANNEL` and `READ_MESSAGE_HISTORY`. Owners and admin users with `ADMINISTRATOR` permissions are exempt.
https://discord.com/developers/docs/topics/permissions#permissions-for-timed-out-members

Author: I think condensing this to "removes all permissions except to read
messages" suffices — `VIEW_CHANNEL` and `READ_MESSAGE_HISTORY` are,
after all, the required combination to read messages.

We could potentially also get away with removing the note about
administrators and owners being exempt from this, as it's clearly stated
that `ADMINISTRATOR` (the permission) overrides anything else.

in the channel a warning was issued in. This can help provide context for past warnings and further inform future mod
actions. This is enabled in the warnings tab on the control panel.

#### Warning Mod Log

You may optionally send warnings to the mod log, this is enabled in the warnings tab on the control panel.

#### Managing Warnings

[Warn Command Syntax](/commands#warn)

[Warnings Command Syntax](/commands#warnings) - Lists warnings of a user.

[EditWarning Command Syntax](/commands#editwarning) - Retroactively edits a warning.

[DelWarning Command Syntax](/commands#delwarning) - Deletes a warning.

[ClearWarnings Command Syntax](/commands#clearwarnings) - Clears the warnings of a user.

[TopWarnings Command Syntax](/commands#topwarnings) - Shows ranked list of warnings on the server, allowing you to turn
misbehavior into a fun minigame for your server members.

{{% notice style="tip" title="Taking Action on Warn Count" %}}

Some users prescribe an "x warnings => mute/kick/ban" rule for their servers. To have YAGPDB automatically take action
based on warning count, you can take advantage of the template scripting within the warning DM.

Example:

```go

> {{/* Config Area Starts */}}
> 
> {{$count := 4 }} {{/* No of warnings at which Certain Action is to be taken ( eg: for action to take place at 4 warnings set count as 4 ) */}}
> {{$some_channel_ID := 0}} {{/* replace with channel id of any random channel which is used for command execution when a default channel of warn doesn't exist */}}
> 
> {{/* Config Area Ends */}}
> 
> {{define "punish_check"}}
> {{$count := toInt .TemplateArgs}}
> {{$warns_count := toInt (reFind `\d+` (execAdmin "Warnings" .User.ID).Description)}}
> 
> {{/* If warning count is above a threshold value take action */}}
> {{if ge $warns_count $count}}
> 
> {{/* Add required code for action to be taken when warning count exceeds the threshold */}}
> Add Code Here
> 
> {{/* Code to reset warnings for the user. Delete if you do not wish to reset the warnings */}}
> {{$silent := execAdmin "ClearWarnings" .User.ID}}
> 
> {{/*End statement for if. Do not Remove if you don't know what you're doing*/}}
> {{end}}
> 
> {{end}}
> 
> {{if .Channel}}
>      {{$s := sendTemplate .Channel.ID "punish_check" $count}}
> {{else}}
>      {{$s := sendTemplate $some_channel_ID "punish_check" $count}}
> {{end}}

Reviewer: If this example is to be part of the official documentation, we should clean it up a little. (Note: untested.)
```suggestion
{{ $threshold := 4 }} {{/* Number of warnings at which action is to be taken (eg: for action to take place at 4 warnings set threshold to 4) */}}

{{ define "punish_check" }}
	{{ $threshold := toInt .TemplateArgs }}
	{{ $warnCount := toInt (reFind `\d+` (execAdmin "warnings" .User).Description) }}

	{{if ge $warns_count $count}}
		{{/* Add required code for action to be taken when warning count exceeds the threshold. */}}
		Add code here

		{{/* Reset warnings for the user. Delete if you do not wish to reset the warnings. */}}
		{{ $silent := execAdmin "clearwarnings" .User }}
	{{ end }}
{{ end }}

{{ if .Channel }}
     {{ $s := sendTemplate .Channel.ID "punish_check" $count }}
{{ else }}
	{{/* no context channel; use an arbitrary one */}}
     {{ $s := sendTemplate (index .Guild.Channels 0).ID "punish_check" $count }}
{{ end }}
```

Reviewer 3: `$count` was only changed to `$threshold` in the definitions, and this should use `getWarnings` instead of parsing an embed from `execAdmin`.
